doc/api/stream.md

```diff
@@ -1555,12 +1555,23 @@ const cleanup = finished(rs, (err) => {
 });
 ```
 
-### `stream.pipeline(...streams, callback)`
+### `stream.pipeline(source, ...transforms, destination, callback)`
 <!-- YAML
 added: v10.0.0
--->
-
-* `...streams` {Stream} Two or more streams to pipe between.
+changes:
+  - version: REPLACEME
+    pr-url: https://github.com/nodejs/node/pull/31223
+    description: Add support for functions and generators.
+-->
+
+* `source` {Stream|Iterable|AsyncIterable|Function}
+  If `Function` then `source` must return a value of `Promise` of
+  `Iterable` or `AsyncIterable`.
+* `...transforms` {Stream|Function}
+  If `Function` then `source` must return a value of `Promise` of `AsyncIterable`.
+  * `source` {AsyncIterable}
+* `destination` {Stream|Function}.
+  * `source` {AsyncIterable}
 * `callback` {Function} Called when the pipeline is fully done.
   * `err` {Error}
 
```

Comment:
Probably needs a `Returns: ` section. Pending some other TODOs/questions below. Especially in regards to whether or not to return promises.